packages/connection/src/Client.php

```diff
@@ -77,64 +96,67 @@ public static function remote_request( $args, $body = null ) {
 				// We cast this to a new variable, because the array form of $body needs to be
 				// maintained so it can be passed into the request later on in the code.
 				if ( count( $body ) > 0 ) {
-					$body_to_hash = json_encode( self::_stringify_data( $body ) );
+					$body_to_hash = wp_json_encode( self::_stringify_data( $body ) );
 				} else {
 					$body_to_hash = '';
 				}
 			}
 
 			if ( ! is_string( $body_to_hash ) ) {
-				return new Jetpack_Error( 'invalid_body', 'Body is malformed.' );
+				return new \WP_Error( 'invalid_body', 'Body is malformed.' );
 			}
 
-			$body_hash = Jetpack::connection()->sha1_base64( $body_to_hash );
+			$body_hash = base64_encode( sha1( $body_to_hash, true ) ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.obfuscation_base64_encode
 		}
 
 		$auth = array(
-			'token' => $token_key,
+			'token'     => $token_key,
 			'timestamp' => $timestamp,
-			'nonce' => $nonce,
+			'nonce'     => $nonce,
 			'body-hash' => $body_hash,
 		);
 
 		if ( false !== strpos( $args['url'], 'xmlrpc.php' ) ) {
 			$url_args = array(
 				'for'           => 'jetpack',
-				'wpcom_blog_id' => Jetpack_Options::get_option( 'id' ),
+				'wpcom_blog_id' => \Jetpack_Options::get_option( 'id' ),
 			);
 		} else {
 			$url_args = array();
 		}
 
-		if ( 'header' != $args['auth_location'] ) {
+		if ( 'header' !== $args['auth_location'] ) {
 			$url_args += $auth;
 		}
 
 		$url = add_query_arg( urlencode_deep( $url_args ), $args['url'] );
-		$url = Jetpack::fix_url_for_bad_hosts( $url );
+		$url = \Jetpack::fix_url_for_bad_hosts( $url );
 
 		$signature = $jetpack_signature->sign_request( $token_key, $timestamp, $nonce, $body_hash, $method, $url, $body, false );
 
-		if ( !$signature || is_wp_error( $signature ) ) {
+		if ( ! $signature || is_wp_error( $signature ) ) {
 			return $signature;
 		}
 
-		// Send an Authorization header so various caches/proxies do the right thing
+		// Send an Authorization header so various caches/proxies do the right thing.
 		$auth['signature'] = $signature;
-		$auth['version'] = JETPACK__VERSION;
-		$header_pieces = array();
+		$auth['version']   = Constants::get_constant( 'JETPACK__VERSION' );
+		$header_pieces     = array();
 		foreach ( $auth as $key => $value ) {
 			$header_pieces[] = sprintf( '%s="%s"', $key, $value );
 		}
-		$request['headers'] = array_merge( $args['headers'], array(
-			'Authorization' => "X_JETPACK " . join( ' ', $header_pieces ),
-		) );
+		$request['headers'] = array_merge(
+			$args['headers'],
+			array(
+				'Authorization' => 'X_JETPACK ' . join( ' ', $header_pieces ),
+			)
+		);
 
-		if ( 'header' != $args['auth_location'] ) {
-			$url = add_query_arg( 'signature', urlencode( $signature ), $url );
+		if ( 'header' !== $args['auth_location'] ) {
+			$url = add_query_arg( 'signature', rawurlencode( $signature ), $url );
```

Comment:
Have we verified that using `rawurlencode` instead of `urlencode` won't cause a trouble?

Reply:
The difference between these two is that `urlencode` turns spaces into plus symbols, and `rawurlencode` turns spaces into proper entities, `%20`. In case of signatures, these are base64 encoded strings that do not have spaces in them. But they can have plus symbols, so in my thinking `rawurlencode` is a better way to go.